File: 123/max_profit.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        num_days = len(prices)
        K = 2
        dp = [[[0, 0] for k in range(K+1)] for i in range(num_days+1)]

        for i in range(num_days+1):
            dp[i][0][0] = 0
            dp[i][0][1] = float('-inf')

        for k in range(K+1):
            dp[0][k][0] = 0
            dp[0][k][1] = float('-inf')

        for i in range(1, num_days+1):
            for k in reversed(range(1, K+1)):
                dp[i][k][0] = max(dp[i-1][k][0], dp[i-1][k][1]+prices[i-1])
                dp[i][k][1] = max(dp[i-1][k][1], dp[i-1][k-1][0]-prices[i-1])

        return int(dp[num_days][K][0])

File: interview/63/max_profit.cpp
#include<vector>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int numDays = prices.size();
        int dp[numDays+1][2];

        dp[0][0] = 0;
        

        for(int i=1; i<=numDays; i++){
            if(i==1){
                dp[i][0] = 0;
                dp[i][1] = -prices[i-1];
            }
            else {
                dp[i][0] = max(dp[i-1][0], dp[i-1][1] + prices[i-1]);
                dp[i][1] = max(dp[i-1][1], - prices[i-1]);
            }
        }

        return dp[numDays][0];
    }
};